lib/workflow/task-graph.js

```diff
@@ -0,0 +1,493 @@
+            if (_.contains(value, Constants.TaskStates.Finished)) {
+                value = Constants.FinishedTaskStates;
+                task.waitingOn[dep] = [Constants.TaskStates.Finished];
+            }
+            // The dependent task is no longer terminal on state <value> since the
+            // current task lists that condition as a dependency.
+            // task.nonTerminalOnStates gets substracted from task.terminalOnStates
+            // when we traverse to the dependent task.
+            nonTerminalOnStates = Array.isArray(value) ? value : [value];
+            return self._visitGraphNode(dep, task.injectableName, markers, nonTerminalOnStates);
+        });
+
+        marker.permanentMark = true;
+        marker.temporaryMark = false;
+    };
+
+    TaskGraph.prototype.detectCyclesAndSetTerminalTasks = function() {
+        var self = this;
+        var markers = {};
+        _.forEach(self.tasks, function(task, taskId) {
+            self._visitGraphNode(taskId, null, markers, null);
+        });
+    };
+
+    /*
+     * Take the tasks definitions in this.definition.tasks, generate instanceIds
+     * to use for each task, and then create new Task objects that reference
+     * the instanceIds in their dependencies instead of user-created task labels.
+     */
+    // TODO: Replace this with a proper DFS traversal instead of iterating
+    TaskGraph.prototype._populateTaskData = function() {
+        var self = this;
+
+        assert.arrayOfObject(self.definition.tasks);
+        var idMap = _.transform(self.definition.tasks, function(result, v) {
+            result[v.label] = uuid.v4();
+        }, {});
+        return Promise.map(self.definition.tasks, function(taskData) {
+            assert.object(taskData);
+            if (_.has(taskData, 'taskName')) {
+                assert.string(taskData.taskName);
+            } else if (_.has(taskData, 'taskDefinition')) {
+                assert.object(taskData.taskDefinition);
+            } else {
+                throw new Error("All TaskGraph tasks should have either a taskName" +
+                    " or taskDefinition property.");
+            }
+
+            _.forEach(_.keys(taskData.waitOn), function(waitOnTask) {
+                var newWaitOnTaskKey = idMap[waitOnTask];
+                assert.ok(newWaitOnTaskKey, 'Task to wait on does not exist: ' + waitOnTask);
+                var waitOnTaskValue = taskData.waitOn[waitOnTask];
+                delete taskData.waitOn[waitOnTask];
+                taskData.waitOn[newWaitOnTaskKey] = waitOnTaskValue;
+            });
+            var taskOverrides = {
+                instanceId: idMap[taskData.label],
+                waitingOn: taskData.waitOn,
+                ignoreFailure: taskData.ignoreFailure
+            };
+
+            if (taskData.taskName) {
+                return self.constructTaskObject(taskData.taskName, taskOverrides,
+                        taskData.optionOverrides, taskData.label)
```

Comment:
From your explanation, this sure brings some benefit to distinguish longer argument or longer logic. I just searched some popular JS style guide, they don't have such a style definition, they always prefer the 4-space indention in nearly every case (the exception case is for argument vertical alignment and `? :`)

What's more, I'm curious how the VIM automatically does this kind of "smart" indentation,  we have to do it manually?